### PR TITLE
add data cell structure, flatten as example

### PR DESCRIPTION
**Pull request by LHT129**
#40

### CODE REVIEW

**Comment by wxyucs** [comment]
@LHT129 please add an issue to describe the "data-cell", and ref it in this pull request description

File: src/data_cell/flatten_datacell.h [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    QueryImpl(float* resultDists, const float* queryVector, const uint64_t* idx, uint64_t idCount);
+
+    inline void
+    QueryImpl(float* resultDists,
+              std::shared_ptr<Computer<QuantTmpl>> computer,
+              const uint64_t* idx,
+              uint64_t idCount);
+
+    inline std::shared_ptr<Computer<QuantTmpl>>
+    FactoryComputerImpl(const float* query) {
```
Comment (ShawnShawnYou) [inline src/data_cell/flatten_datacell.h]: The private method should be written in lowercase with underscores.

Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: change to lowercase

File: src/quantization/computer.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    inline void
+    ComputeDist(const uint8_t* codes, float* dists) {
+        quantizer_->ComputeDist(*this, codes, dists);
+    }
+
+public:
+    const T* quantizer_{nullptr};
+
```
Comment (wxyucs) [inline src/quantization/computer.h]: delete the blank line

File: src/quantization/computer.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    uint8_t* buf_{nullptr};  // todo how to alloc and free
```
Comment (wxyucs) [inline src/quantization/computer.h]: TODO(your id): what to do

File: src/data_cell/flatten_interface.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "quantization/computer.h"
+namespace vsag {
```
Comment (wxyucs) [inline src/data_cell/flatten_interface.h:20]: add a blank line between include and namespace line

File: src/data_cell/flatten_interface.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+          const uint64_t* idx,
+          uint64_t idCount) = 0;
+
+    virtual std::shared_ptr<ComputerInterface>
+    FactoryComputer(const float* query) = 0;
+
+public:
+    virtual float
+    ComputePairVectors(uint64_t id1, uint64_t id2) {
+        return 0.;
```
Comment (wxyucs) [inline src/data_cell/flatten_interface.h]: return 0.0f;

File: src/data_cell/flatten_interface.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    [[nodiscard]] virtual const uint8_t*
+    GetCodesById(uint64_t id) const {
+        return nullptr;
+    }
+
+    [[nodiscard]] virtual uint64_t
+    TotalCount() const {
+        return 0;
+    }
```
Comment (wxyucs) [inline src/data_cell/flatten_interface.h]: what is the basic impl of these functions

File: src/data_cell/flatten_datacell.h [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+#include "quantization/quantizer.h"
+
+namespace vsag {
+
+template <typename QuantTmpl, typename IOTmpl>
+class FlattenDataCell : public FlattenInterface {
+public:
+    FlattenDataCell() = default;
+
+    explicit FlattenDataCell(const std::string& initializeJson);  // todo
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h]: TODO(your id): what to do

File: src/data_cell/flatten_datacell.h [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+public:
+    uint64_t total_count_{0};
+
+    uint64_t max_capacity_{1000000};
+
+    uint64_t code_size_{0};
+
+    std::shared_ptr<Quantizer<QuantTmpl>> quantizer_{nullptr};
+
+    std::shared_ptr<BasicIO<IOTmpl>> io_{nullptr};
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h]: what is the reason to make these variables public? 

remove blank lines, we only add a blank line to distinguish the usage of variables.

File: src/data_cell/flatten_datacell.h [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+private:
+    inline void
+    queryImpl(float* resultDists, const float* queryVector, const uint64_t* idx, uint64_t idCount);
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h]: queryImpl() -> query_impl(), 
queryVector -> query_vector
idCount -> id_count

File: src/data_cell/flatten_datacell.h [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+    inline void
+    queryImpl(float* resultDists,
+              std::shared_ptr<Computer<QuantTmpl>> computer,
+              const uint64_t* idx,
+              uint64_t idCount);
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h]: queryImpl -> query_impl
resultDists -> result_dists
idCount -> id_count

File: src/data_cell/flatten_datacell.h [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+    inline std::shared_ptr<ComputerInterface>
+    factoryComputerImpl(const float* query) {
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h]: factoryComputerImpl -> factory_computer_impl

File: src/data_cell/flatten_datacell.h [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::InsertVector(const float* vector, uint64_t idx) {
+    if (idx == std::numeric_limits<uint64_t>::max()) {
+        idx = total_count_;
+        ++total_count_;
+    }
+    auto* codes = new uint8_t[code_size_];
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h:125]: where is the place to process the alloc failure?

File: src/data_cell/flatten_datacell.h [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+    delete[] codes;
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::BatchInsertVector(const float* vectors,
+                                                      uint64_t count,
+                                                      uint64_t* idx) {
+    if (idx == nullptr) {
+        auto* codes = new uint8_t[code_size_ * count];
```
Comment (wxyucs) [inline src/data_cell/flatten_datacell.h:137]: where is the place to process the alloc failure?

**Review by wxyucs — CHANGES_REQUESTED** [review]

**Comment by wxyucs** [comment]
@LHT129 I left some comments, please have a check

File: src/data_cell/flatten_datacell.h [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    computer->SetQuery(queryVector);
+    this->Query(resultDists, computer, idx, idCount);
+}
+
+template <typename QuantTmpl, typename IOTmpl>
+void
+FlattenDataCell<QuantTmpl, IOTmpl>::queryImpl(float* resultDists,
+                                              std::shared_ptr<Computer<QuantTmpl>> computer,
+                                              const uint64_t* idx,
+                                              uint64_t idCount) {
```
Comment (jiaweizone) [inline src/data_cell/flatten_datacell.h]: idCount -> id_count

File: src/quantization/computer.h (hunk 2 above)
Comment (LHT129) [reply src/quantization/computer.h]: done

File: src/data_cell/flatten_datacell.h (hunk 14 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

File: src/data_cell/flatten_datacell.h (hunk 9 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

File: src/data_cell/flatten_datacell.h (hunk 10 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

File: src/data_cell/flatten_datacell.h (hunk 11 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h]: done

File: src/data_cell/flatten_datacell.h (hunk 12 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h:125]: we will use allocator to alloc and free in future

File: src/data_cell/flatten_datacell.h (hunk 13 above)
Comment (LHT129) [reply src/data_cell/flatten_datacell.h:137]: we will use allocator to alloc and free in future

File: src/data_cell/flatten_interface.h (hunk 6 above)
Comment (LHT129) [reply src/data_cell/flatten_interface.h]: some fixed

**Review by wxyucs — APPROVED** [review]
lgtm

File: src/quantization/computer.h [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+template <typename T>
+class Computer : public ComputerInterface {
+public:
+    ~Computer() {
+        delete[] buf_;
+    }
+
+    explicit Computer(const T& quantizer) : quantizer_(&quantizer){};
```
Comment (jiaweizone) [inline src/quantization/computer.h]: There can replace to `const T*` ?
The const reference is not safe, example:

```C++
#include <iostream>

const int* ptr = nullptr;

void test(const int& ref) {
    ptr = &ref;
    std::cout << *ptr << std::endl;
}

void test2() {
    int num = 2;
    test(1);
    test(num);
}

int main(int argc, char* argv[]) {

    test2();
    std::cout << *ptr << std::endl; // access undefined memory

    return 0;
}
```

Comment (LHT129) [reply src/quantization/computer.h]: fixed, use const T* instead

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**